samples/300/connect_retry_with_telemetry.py

```diff
@@ -0,0 +1,168 @@
+# To have a more detailed experience
+transient_errors = [
+    exceptions.OperationCancelled,
+    exceptions.OperationTimeout,
+    exceptions.ServiceError,
+    exceptions.ConnectionFailedError,
+    exceptions.ConnectionDroppedError,
+    exceptions.NoConnectionError,
+    exceptions.ClientError,  # TODO Only TLSEsxhangeError is actually worth retrying
+]
+
+
+def is_retryable(exc):
+    if type(exc) in transient_errors:
+        return True
+    else:
+        return False
+
+
+def create_client():
+    global connected_event
+    global disconnected_event
+
+    # Create a Device Client
+    device_client = IoTHubDeviceClient.create_from_connection_string(CONNECTION_STRING)
+
+    connected_event = asyncio.Event()
+    disconnected_event = asyncio.Event()
+
+    # Define a connection state change handler
+    async def handle_on_connection_state_change():
+        global main_event_loop
+        print(
+            "handle_on_connection_state_change fired. Connected status : {}".format(
+                device_client.connected
+            )
+        )
+        if device_client.connected:
+            print("Connected event is set...")
+            main_event_loop.call_soon_threadsafe(connected_event.set)
+        else:
+            print("Disconnected event is set...")
+            main_event_loop.call_soon_threadsafe(disconnected_event.set)
+
+    try:
+        # Attach the connection state handler
+        device_client.on_connection_state_change = handle_on_connection_state_change
+    except Exception:
+        # Clean up in the event of failure
+        device_client.shutdown()
+        raise
+
+    return device_client
+
+
+async def connect_with_retry(device_client, number_of_tries=None):
+    if device_client.connected:
+        print("Client is already connected...")
+        return True
+    for i in range(number_of_tries):
+        try:
+            print("Attempting to connect the device client try number {}....".format(i))
+            await device_client.connect()
+            print("Successfully connected the device client...")
+            return True
+        except Exception as e:
+            print("Caught exception while trying to connect...")
+            if type(e) is exceptions.CredentialError:
+                print(
+                    "Failed to connect the device client due to incorrect or badly formatted credentials..."
+                )
+                return False
+
+            if is_retryable(e):
+                print(
+                    "Failed to connect the device client due to retryable error.Sleeping and retrying after some time..."
+                )
+                await asyncio.sleep(ATTEMPT_INTERVAL)
+            else:
+                print("Failed to connect the device client due to not-retryable error....")
+                return False
+
+
+async def run_sample(device_client):
+    encountered_no_error = await connect_with_retry(device_client, RETRY_NOS)
```

Comment:
Why is there a max retry on the initial connect anyway? It seems inconsistent - if the application fails too many times while trying to connect, it throws a RuntimeError, but once that first connect happens, the network can drop indefinitely without application exit.